### ISSUE TITLE
No alpha channel support?

### ISSUE THREAD

**Issue by seanocali** [Microsoft/FFmpegInterop#201]
I tried this out with some videos that have transparent backgrounds.  A .webm vp9a, and a .flv vp6a.  Both videos play but the alpha channel is ignored.  I tried playing to a MediaPlayerElement and I also tried playing it to a Win2d canvas via frameserver.  Really not sure why the latter wouldn't work. This article has a link to one of the sample videos I used. [](https://developers.google.com/web/updates/2013/07/Alpha-transparency-in-Chrome-video)

**Comment by lukasf** [comment]
Nope, FFmpegInterop does not support alpha channel. I am pretty sure that MediaPlayerElement also does not support alpha channel. It has a kind of hard coded black background which cannot be disabled or changed afaik. So even if we supported an alpha channel format, I assume that MediaPlayerElement would just draw the transparent areas black.

**Comment by lukasf** [comment]
If I have some time, I might be able to check if alpha support would work with MediaPlayerElement. But you did not add the link to your article. Currently, FFmpegInterop is hard coded to NV12 format (which does not have an alpha channel), so alpha is dropped during decoding. It should be possible to change to ARGB32, but as I said, I am not sure if this will help...

**Comment by seanocali** [comment]
Thanks.  I put the link in to the webm videos.

And ARGB32 output might still be useful.  On the last section of this page:
[Use MediaPlayer in frame server mode](https://docs.microsoft.com/en-us/windows/uwp/audio-video-camera/play-audio-and-video-with-mediaplayer)

It's a way to play a video on a Win2d canvas (which does support alpha) instead of a MediaPlayerElement.  It converts the video frames to CanvasBitmap objects and can display them sequentially like a video.  
[https://microsoft.github.io/Win2D/html/T_Microsoft_Graphics_Canvas_CanvasBitmap.htm](https://microsoft.github.io/Win2D/html/T_Microsoft_Graphics_Canvas_CanvasBitmap.htm)

**Comment by lukasf** [comment]
I've done some experiments with this, but failed to get it working. It looks like ffmpeg does not support alpha channel with webm video. Webm uses VP8 codec which was never designed to include alpha channel. The way to get alpha into webm is pretty hacky and it looks like ffmpeg does not support that. The VP8 decoder is hard coded to YUV420P output format which does not have alpha.

Please check this page, where different hacks for webm alpha are discussed. Not sure which hack was finally used by google, but it can't be a clean solution:

http://wiki.webmproject.org/alpha-channel

In this ffmpeg discussion (different topic) it is also noted that the ffmpeg VP8 decoder does not support alpha channel:

https://patchwork.ffmpeg.org/patch/2607/

Sorry, looks like you are out of luck. Except if you find a different format that officially allows alpha channel and is supported by ffmpeg.

**Comment by seanocali** [comment]
Not sure what ffmpeg version your project uses but newer versions support it:
[https://stackoverflow.com/questions/22485022/splitting-webm-video-to-png-with-transparency](https://stackoverflow.com/questions/22485022/splitting-webm-video-to-png-with-transparency)

I just ran it myself with a regular build from 2017 and it works fine.  Not a custom build just the standard Windows build ffmpeg links to from their site.

![img-0026](https://user-images.githubusercontent.com/16900871/34321098-d788c32c-e7bb-11e7-869d-56c5456ac88e.png)


Another format is Flash video VP6.  There is a link to a sample video here:
[https://www.adobe.com/support/documentation/en/flash/fl8/samples.html](https://www.adobe.com/support/documentation/en/flash/fl8/samples.html)

ffmpeg decodes that format with alpha channel without any arguments.  "ffmpeg -i alpha_vid.flv image-%03d.png" works fine.

Thanks again for looking into this.

**Comment by lukasf** [comment]
I have some good and some bad news for you. First the bad news: It looks like ffmpeg supports webm-alpha only if libvpx is included in the build. This is not the default and it is not currently enabled in WinRT builds used by FFmpegInterop. I am not sure if libvpx is even compatible with WinRT, and it requires libvorbis und libogg to be linked as well (both are not linked right now, but I believe they work with WinRT). This is something that could be possible, but I don't have time to try that right now. If you need webm-alpha, you'd probably have to look into this yourself.

But here comes the good news: Playing the flash file with alpha channel worked well, even with MediaPlayerElement! Strange thing is: MediaPlayerElement still draws black borders on top and bottom of video, but inside the actual video area, the app background is visible through transparent parts of the video (I changed background to blue, and can see that in the video). If you resize the MediaPlayerElement to match the video area and/or set video to "stretch", you will get full alpha support. Too bad that you cannot just change/disable the black MediaPlayerElement background color. But still, you can get it to work somehow. Or you use your Win2d approach, which should work as well.

Please check out my branch "lukasf/alpha-support". Currently, I just hard coded things to ARGB output, which obviously needs to be changed if things are supposed to come officially into FFmpegInterop. Not sure how that will look like. But you can test and play already with the branch. And if you manage to build ffmpeg with libvpx enabled, your webm files should work as well.

**Comment by seanocali** [comment]
Nice I'll check it out!  Well done.  For my own selfish purposes on my project I actually only care about the Flash format so this is great news for me.

**Comment by seanocali** [comment]
So I got your branch working on my project and it plays the flv sample I sent you with transparent background like you said.  Unfortunately all the flv files I was hoping to use (trying to support an older application's user-generated content) do not work, they only play audio.  Same thing when opening them in your MediaPlayerCS.  Here is one example:
[https://drive.google.com/file/d/1MP7XxtkGbQWJLw1e3cOjSczK5eTRhZYa/view?usp=sharing](https://drive.google.com/file/d/1MP7XxtkGbQWJLw1e3cOjSczK5eTRhZYa/view?usp=sharing)

It's very strange ffmpeg handles them fine, VLC plays them fine.  They're the same vp6a video codec.  I tried stripping out the audio stream.  I replaced the metadata to match.  Nothing worked.

I don't expect you help much regarding this but I figured I'd give feedback that apparently not all flv vp6a videos will work.  Also as you suspected I could not figure out a way to build libvpx for WinRT.

**Comment by reego-fr** [comment]
I cant promise you anything but I will have a look at these files and see if something can be done to let them play

**Comment by reego-fr** [comment]
There was actually an issue with FFmpegInterop on playback start with your file. https://github.com/Microsoft/FFmpegInterop/tree/gillesk/SamplePresentationTimestamp solves it.

**Comment by seanocali** [comment]
Thanks man!  Is that based on master branch or the alpha channel one?  If it's master no worries I can just copy and paste the changes.

**Comment by lukasf** [comment]
Well done @reego-fr. We really need to merge all these fixes into master branch soon, to have a clean state again.

**Comment by lukasf** [comment]
@seanocali You can try to use this in OnStarting():
```
			int64_t streamStartTime = avFormatCtx->streams[streamIndex]->start_time == AV_NOPTS_VALUE ? (int64_t)0 : avFormatCtx->streams[streamIndex]->start_time;
			int64_t seekTarget = static_cast<int64_t>(request->StartPosition->Value.Duration / (av_q2d(avFormatCtx->streams[streamIndex]->time_base) * 10000000));

			if (av_seek_frame(avFormatCtx, streamIndex, seekTarget + streamStartTime, AVSEEK_FLAG_BACKWARD) < 0)
```

Not 100% sure if this will work though. Otherwise, you'll have to wait until those PRs are all merged into master branch.

**Comment by seanocali** [comment]
lukas, adding those lines worked BTW.  Most of the time playback is perfect, occasionally it stutters or freezes, but it's more than enough to give me confidence to go forward with the alpha channel feature knowing a future ffmpeginterop release will work great come time for production.  Thanks again to you and reego for making this happen.
